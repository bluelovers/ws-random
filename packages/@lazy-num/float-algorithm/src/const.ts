
export const MATH_POW_2_32 = Math.pow(2, 32);
