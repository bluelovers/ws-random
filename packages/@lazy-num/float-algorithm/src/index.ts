export * from './algorithm/number/mulberry32';
export * from './algorithm/number/splitmix32';
export * from './algorithm/int-list/sfc32';
export * from './algorithm/int-list/tyche';
