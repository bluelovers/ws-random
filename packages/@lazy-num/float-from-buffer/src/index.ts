import { FLOAT_ENTROPY_BYTES, MATH_POW_2_32 } from '@lazy-random/shared-lib';

/**
 * Given a buffer containing bytes of entropy, generate a double-precision
 * 64-bit float.
 *
 * @param {Buffer} buf a buffer of bytes
 * @returns {Number} a float
 *
 * @see https://github.com/fardog/node-random-lib/blob/master/index.js
 * @see http://stackoverflow.com/questions/15753019/floating-point-number-from-crypto-randombytes-in-javascript
 */
export function floatFromBuffer(buf: ArrayLike<number>, offset: number = 0)
{
	offset = Math.floor(offset);

	if (buf.length < (FLOAT_ENTROPY_BYTES + offset) || offset < 0)
	{
		throw new RangeError(
			`buffer must contain at least ${FLOAT_ENTROPY_BYTES}${offset > 0 ? ' +' + offset : ''} bytes of entropy`,
		)
	}

	return _floatFromBuffer(buf, offset)
}

export function _floatFromBuffer(buf: ArrayLike<number>, offset: number = 0)
{
	let position = Math.floor(offset);

	return (((((((
								buf[position++] % 32) / 32 +
							buf[position++]) / 256 +
						buf[position++]) / 256 +
					buf[position++]) / 256 +
				buf[position++]) / 256 +
			buf[position++]) / 256 +
		buf[position]) / 256
}

export function _floatFromBuffer2(buf: ArrayLike<number>, offset: number = 0)
{
	return readUInt32BE(buf, offset) / MATH_POW_2_32
}

export function readUInt32LE(buf: ArrayLike<number>, offset: number = 0)
{
	offset = offset >>> 0;

	return ((buf[offset]) |
			(buf[offset + 1] << 8) |
			(buf[offset + 2] << 16)) +
		(buf[offset + 3] * 0x1000000)
}

export function readUInt32BE(buf: ArrayLike<number>, offset: number = 0)
{
	offset = offset >>> 0;

	return (buf[offset] * 0x1000000) +
		((buf[offset + 1] << 16) |
			(buf[offset + 2] << 8) |
			buf[offset + 3])
}

export default floatFromBuffer
