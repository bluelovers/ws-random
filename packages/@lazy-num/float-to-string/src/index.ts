export function assertFractionDigits(fractionDigits?: number): asserts fractionDigits is number
{
	if (!Number.isFinite(fractionDigits) || !Number.isInteger(fractionDigits) || fractionDigits === 0)
	{
		throw new TypeError(`Invalid fractionDigits: ${fractionDigits}`)
	}
}

export function splitFloatNumberToString(float: string | number): [string, string?]
{
	return String(float).split('.') as any
}

export function getFractionDigitsString(float: string | number)
{
	return splitFloatNumberToString(float)[1]
}

export function splitFloatNumber(n: number): [number, number]
{
	let int = Math.floor(n);
	let float = n - int;

	return [int, float]
}

export function joinFloatNumber(int: number | string, float?: string)
{
	return String(int) + (float?.length ? '.' + float : '');
}

export function floatToString(n: number, fractionDigits?: number)
{
	let int: string | number;
	let s: string;

	if (typeof fractionDigits === 'number')
	{
		assertFractionDigits(fractionDigits);

		([int, s] = splitFloatNumberToString(n.toFixed(fractionDigits)));
	}
	else
	{
		let float;
		([int, float] = splitFloatNumber(n));
		s = getFractionDigitsString(float);
	}

	return joinFloatNumber(int, s);
}

export default floatToString
