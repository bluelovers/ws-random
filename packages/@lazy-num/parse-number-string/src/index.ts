
const reInt = /^[+-]?\d+$/;
const reFloatOnly = /^[+-]?(?:\d+)?\.\d+$/;
const reFloat = new RegExp(reInt.source + '|' + reFloatOnly.source);

export function isIntString<T extends `${number}` | string = `${number}`>(input: unknown): input is T
{
	return (typeof input === 'string' && reInt.test(input))
}

export function isFloatOnlyString<T extends `${number}` | string = `${number}`>(input: unknown | T): input is T
{
	return (typeof input === 'string' && reFloatOnly.test(input))
}

export function isFloatString<T extends `${number}` | string = `${number}`>(input: unknown | T): input is T
{
	return (typeof input === 'string' && reFloat.test(input))
}

export type IParseNumberFn = (<T extends `${number}` | string = `${number}`>(input: unknown | T) => input is T) | ((input: unknown) => input is`${number}`)

export function _parseNumberString<T extends number>(validFn: IParseNumberFn, input: unknown | T | `${T}`): T
{
	if (validFn(input))
	{
		return Number(input) as T
	}

	if (typeof input !== 'number')
	{
		throw new TypeError(`Invalid value: ${input}`)
	}

	return input as T
}

export function parseIntString<T extends number>(input: T): T
export function parseIntString<T extends number>(input: `${T}`): T
export function parseIntString<T extends number>(input: unknown | T | `${T}`): T
export function parseIntString<T extends number>(input: unknown | T | `${T}`): T
{
	return _parseNumberString(isIntString, input)
}

export function parseFloatOnlyString<T extends number>(input: T): T
export function parseFloatOnlyString<T extends number>(input: `${T}`): T
export function parseFloatOnlyString<T extends number>(input: unknown | T | `${T}`): T
export function parseFloatOnlyString<T extends number>(input: unknown | T | `${T}`): T
{
	return _parseNumberString(isFloatOnlyString, input)
}

export function parseFloatString<T extends number>(input: T): T
export function parseFloatString<T extends number>(input: `${T}`): T
export function parseFloatString<T extends number>(input: unknown | T | `${T}`): T
export function parseFloatString<T extends number>(input: unknown | T | `${T}`): T
{
	return _parseNumberString(isFloatString, input as any)
}

export default parseFloatString
