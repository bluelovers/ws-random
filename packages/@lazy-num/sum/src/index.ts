import { fixZero } from 'num-is-zero';

/**
 * (1 + 2 + 3 +...+N)
 *
 * @see http://emn178.pixnet.net/blog/post/92132837-%E8%A8%88%E7%AE%971%E5%88%B0n%E7%B8%BD%E5%92%8C%281-%2B-2-%2B-3-%2B...%2Bn%29
 */
export function sum_1_to_n(n: number)
{
	return n * (n + 1) / 2;
}

export function num_array_sum(na: number[])
{
	return fixZero(na.reduce((a, b) => a + b))
}
