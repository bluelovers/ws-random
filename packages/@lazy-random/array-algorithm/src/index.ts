
export * from './swapAlgorithm';
export * from './array_rebase';
