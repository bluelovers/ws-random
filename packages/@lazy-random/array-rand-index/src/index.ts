import { _MathRandom } from '@lazy-random/original-math-random';

export function arrayRandIndexByLength(len: number, ...argv: any[])
{
	return Math.floor(_MathRandom() * len)
}

export function arrayRandIndex(array: any[], ...argv: any[])
{
	return arrayRandIndexByLength(array.length)
}

export default arrayRandIndex
