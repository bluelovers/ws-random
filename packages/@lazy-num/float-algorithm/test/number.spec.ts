import { df_mulberry32, df_splitmix32 } from '../src/index';
import { df_xfnv1a } from '@lazy-random/seed-algorithm';

describe(`number`, () =>
{
	const _t1 = df_xfnv1a('apple')();
	const _t2 = df_xfnv1a('applE')();

	([
		df_mulberry32,
		df_splitmix32,
	]).forEach(fn =>
	{

		test(fn.name, () =>
		{

			const seedFn: () => number = fn(_t1)

			let actual = [
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
			];

			expect(actual).toMatchSnapshot();

			const seedFn2: () => number = fn(_t2)

			let actual2 = [
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
			];

			expect(actual).not.toStrictEqual(actual2);

			expect(actual2).toMatchSnapshot();

		});

	})

})
