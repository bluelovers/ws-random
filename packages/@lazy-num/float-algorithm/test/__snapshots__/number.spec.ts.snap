// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`number df_mulberry32 1`] = `
[
  0.07845377526246011,
  0.44624303048476577,
  0.2832711616065353,
  0.7845816505141556,
  0.9755276739597321,
  0.44098455156199634,
  0.06808405858464539,
  0.3188576155807823,
  0.6483108345419168,
  0.1686609056778252,
  0.060697185806930065,
  0.7204918644856662,
  0.22251161909662187,
  0.19648338085971773,
  0.39632478170096874,
  0.6545514219906181,
]
`;

exports[`number df_mulberry32 2`] = `
[
  0.7439509234391153,
  0.6870627191383392,
  0.4839799259789288,
  0.8153954423032701,
  0.6879334182012826,
  0.9649797754827887,
  0.666028386214748,
  0.7311498147901148,
  0.21718146419152617,
  0.8067912785336375,
  0.5702531803399324,
  0.798434310592711,
  0.6636752306949347,
  0.7637704347725958,
  0.5261733690276742,
  0.7091079936362803,
]
`;

exports[`number df_splitmix32 1`] = `
[
  0.4224637313745916,
  0.47743686009198427,
  0.901220950530842,
  0.21907008439302444,
  0.21955096209421754,
  0.13954805629327893,
  0.9507406358607113,
  0.1713115160819143,
  0.7375910996925086,
  0.8956481541972607,
  0.988201015163213,
  0.45239880913868546,
  0.04964208975434303,
  0.8978918592911214,
  0.6366649500560015,
  0.8664729951415211,
]
`;

exports[`number df_splitmix32 2`] = `
[
  0.6433862149715424,
  0.9569811569526792,
  0.3575705736875534,
  0.7158022362273186,
  0.05328744649887085,
  0.47713084309361875,
  0.12533995276317,
  0.24278043489903212,
  0.5424668625928462,
  0.13108001369982958,
  0.8964921156875789,
  0.7554246606305242,
  0.6246357860509306,
  0.04523041471838951,
  0.023794775130227208,
  0.5285730082541704,
]
`;
